File: exp4/check_answer.py
# -*- coding: utf-8 -*-

def load_file(file_path):
    with open(file_path, 'r') as f:
        return f.readlines()

def compare_str(line1, line2):
    str1, str2 = line1.strip(), line2.strip()
    return str1 == str2

def compare(result_path, compare_path):
    result_lines = load_file(result_path)
    compare_lines = load_file(compare_path)
    if len(result_lines) != len(compare_lines):
        print("check failed with unequal lines")
        print("origin lines:", len(result_lines))
        print("compare lines: ", len(compare_lines))
        return
    for line1, line2 in zip(result_lines, compare_lines):
        if not compare_str(line1, line2):
            print("check failed...")
            print("origin output is:", line1)
            print("compare output is: ", line2)
            return
    print("check passed")

if __name__ == '__main__':
    result_path = './exp7_out.txt'
    compare_path = './myexp7_out.txt'
    compare(result_path, compare_path)

File: exp3/check_answer.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <cstring>
using namespace std;

vector<string> load_file(string file_path) {
    vector<string> result;
    freopen(file_path.c_str(), "r", stdin);
    string temp;
    while (getline(cin, temp)) {
        result.push_back(temp);
    }
    fclose(stdin);
    cin.clear();
    return result;
}

bool compare_str(string str1, string str2) {
    int end1 = str1.size() - 1, end2 = str2.size() - 1;
    while (end1 > 0 && (str1[end1] == ' ' || str1[end1] == '\t')) {
        end1--;
    }
    while (end2 > 0 && (str2[end2] == ' ' || str2[end2] == '\t')) {
        end2--;
    }
    if (end1 != end2) return false;
    for (int i = 0; i <= end1; i++) {
        if (str1[i] != str2[i]) {
            return false;
        }
    }
    return true;
}

void compare(string result_path, string compare_path) {
    vector<string> result_lines = load_file(result_path);
    vector<string> compare_lines = load_file(compare_path);
    int i = 0, j = 0, len1 = result_lines.size(), len2 = compare_lines.size();
    if (len1 != len2) {
        cout << "check failed with unequal lines\n";
        cout << "origin lines: " << len1 << endl;
        cout << "compare lines: " << len2 << endl;
        return ;
    }
    for (int i = 0; i < len1; i++) {
        if (!compare_str(result_lines[i], compare_lines[i])) {
            cout << "check failed...\n";
            cout << "origin output is: " << result_lines[i] << endl;
            cout << "compare output is: " << compare_lines[i] << endl;
            return ;
        }
    }
    cout << "check passed\n";
}

int main() {
    string result_path = "./exp3_out.txt";
    string compare_path = "./exp3_out1.txt";
    compare(result_path, compare_path);
    
    return 0;
}
